content/en/blog/2021-12-06-fake-it-when-you-make-it.md

```diff
@@ -0,0 +1,266 @@
+		c.JSON(http.StatusBadRequest, err)
+		return
+	}
+	if access, err = generateToken(); err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+	if refresh, err = generateToken(); err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+	c.JSON(http.StatusOK, &sectigo.AuthenticationReply{
+		AccessToken:  access,
+		RefreshToken: refresh,
+	})
+}
+```
+
+There are a few things going on here. We are using `httptest` and `gin` to quickly spin up a test server and configure an endpoint to our mocked `authenticate` handler. We also call the project-specific method `sectigo.SetBaseURL` to ensure that all the Sectigo requests get directed to our mock. In the mocked `authenticate` handler, we simply generate some valid-looking JWT tokens and send them back in the request so our Sectigo client code hits the "success" path. Finally, we have a call counter which helps verify that the correct endpoints are being called. This infrastructure enables us to write a simple test[^4] for our Sectigo client `Authenticate` method.
+
+```go
+func TestAuthenticate(t *testing.T) {
+    m, err := mock.New()
+    require.NoError(t, err)
+    defer m.Close()
+    s.api, err = NewSectigoClient("foo", "supersecret")
+    require.NoError(t, err)
+	require.NoError(t, s.api.Authenticate())
+}
+```
+
+What happens if we run into an edge case we want to add a test for? With the mock we've set up, this is actually pretty straightforward. For example, if we want to test that our Sectigo client is handling invalid credentials properly, we can replace the mock `authenticate` handler with one that has validation logic.
+
+```go
+func TestAuthenticateInvalidCreds() {
+	m, err := mock.New()
+	require.NoError(t, err)
+	defer m.Close()
+
+	m.Handle(AuthenticateEP, func(c *gin.Context) {
+		var (
+			in *AuthenticationRequest
+		)
+		if err := c.BindJSON(&in); err != nil {
+			c.JSON(http.StatusBadRequest, err)
+			return
+		}
+		if in.Username != "foo" || in.Password != "supersecret" {
+			c.JSON(http.StatusUnauthorized, "invalid credentials")
+			return
+		}
+
+		c.JSON(http.StatusInternalServerError, "how did we get here?")
+	})
+
+	s.api, err = New("invalid", "invalid")
+	require.NoError(t, err)
+	err = s.api.Authenticate()
+	require.EqualError(t, err, ErrInvalidCredentials.Error())
+}
+```
+
+## Conclusion
+
+Testing distributed systems applications is difficult due to the many interdependencies involved and the design of programming languages such as Go. In this post, we've looked at a few examples of how mocks can be used to alleviate some of these pain points.
+
+Mocks can help us bypass code that we don't want to run in our tests without fundamentally changing the architecture of our applications. With some configuration, mocks allow us to test both the success and error paths of our service-dependent code, enabling more complete unit testing, and a greater degree of confidence in the implemention of highly complex systems.
```

Comment:
I added an additional paragraph break here and a few additional words to the end of the last sentence to bring us back to distsys -- hope that's ok!

Reply:
That sounds great to me!